File: server/request.py
import parser
from request_payloads import NamePayload, KeyPayload, FilePayload, CRCPayload
from consts import Sizes, RequestCode


#  Request parsing obj
class Request:

    def __init__(self, bytes_request: bytearray):
        request_index = 0
        # set client id from byte data
        self.client_id = parser.get_str_from_bytes(byte_arr=bytes_request, from_ind=request_index,
                                                   size=Sizes.CLIENT_ID_SIZE)
        request_index += Sizes.CLIENT_ID_SIZE
        # set version from byte data
        self.version = parser.get_int_from_bytes(byte_arr=bytes_request, from_ind=request_index,
                                                 size=Sizes.VERSION_SIZE)
        request_index += Sizes.VERSION_SIZE
        # set request code from byte data
        self.code = parser.get_int_from_bytes(byte_arr=bytes_request, from_ind=request_index,
                                              size=Sizes.CODE_SIZE)
        request_index += Sizes.CODE_SIZE
        # set payload len from byte data
        self.payload_len = parser.get_int_from_bytes(byte_arr=bytes_request, from_ind=request_index,
                                                     size=Sizes.PAYLOAD_LEN_SIZE)
        request_index += Sizes.PAYLOAD_LEN_SIZE
        bytes_payload = parser.get_sub_byte_arr(byte_arr=bytes_request, from_ind=request_index,
                                                size=self.payload_len)
        # set payload obj from byte data
        match self.code:
            case RequestCode.REGISTRATION:
                self.payload = NamePayload(bytes_payload=bytes_payload)
            case RequestCode.SEND_PUBLIC_KEY:
                self.payload = KeyPayload(bytes_payload=bytes_payload)
            case RequestCode.SEND_FILE:
                self.payload = FilePayload(bytes_payload=bytes_payload)
            case RequestCode.CRC_OK | RequestCode.CRC_SECOND | RequestCode.CRC_ERROR:
                self.payload = CRCPayload(bytes_payload=bytes_payload)
            case _:
                raise Exception("Invalid Request code. (The request content is incorrect)")

File: client/Request.cpp
#include "Request.h"

// build the request obj
Request::Request(char* cid, int code, char* payload, int payload_len) {
	this->data_len = CLIENT_ID_SIZE + VERSION_SIZE + CODE_SIZE + PAYLOAD_LEN_SIZE + payload_len + 1;
	// build the bytes request
	this->data = new char[data_len];
	for (int i = 0; i < this->data_len; i++)
		this->data[i] = '\0';
	if (cid)
		memcpy(this->data, cid, CLIENT_ID_SIZE);
	int version = 3;
	memcpy(this->data + CLIENT_ID_SIZE, &version, VERSION_SIZE);
	memcpy(this->data + CLIENT_ID_SIZE + VERSION_SIZE, &code, CODE_SIZE);
	memcpy(this->data + CLIENT_ID_SIZE + VERSION_SIZE + CODE_SIZE, &payload_len, PAYLOAD_LEN_SIZE);
	memcpy(this->data + CLIENT_ID_SIZE + VERSION_SIZE + CODE_SIZE + PAYLOAD_LEN_SIZE, payload, payload_len);
}

Request::~Request() {
	delete this->data;
}

char* Request::getRequestData() {
	return data;
}

int Request::get_data_len(){
	return this->data_len;
}

Registration::Registration(char* name) : Request(NULL, REGISTRATION, name, NAME_SIZE) {}

PublicKeyReq::PublicKeyReq(char* cid, char* payload) : Request(cid, SEND_PUBLIC_KEY, payload, FILE_NAME_SIZE + PUBLIC_KEY_SIZE) {}

SendFileReq::SendFileReq(char* cid, char* payload, int size) : Request(cid, SEND_FILE, payload, CLIENT_ID_SIZE + CONTENT_SIZE + FILE_NAME_SIZE + size) {}

CrcOk::CrcOk(char* cid, char* payload) : Request(cid, CRC_OK, payload, CLIENT_ID_SIZE + FILE_NAME_SIZE) {}

CrcSendAgein::CrcSendAgein(char* cid, char* payload) : Request(cid, CRC_SECOND, payload, CLIENT_ID_SIZE + FILE_NAME_SIZE) {}

CrcFail::CrcFail(char* cid, char* payload) : Request(cid, CRC_ERROR, payload, CLIENT_ID_SIZE + FILE_NAME_SIZE) {}
